### PR TITLE
[SYCL] Add E2E tests for assert feature

### PR DESCRIPTION
**Pull request by alexanderfle**
Created 9 E2E tests for assert feature in accordance with the requirements:

Preliminary list of use-cases:
- single kernel in application
- multiple kernels in application, one kernel working at a time
- multiple kernels in application, multiple kernels working simultaneously

Orthogonal to this list, compilation use-cases list is the following:
- application created from single translation unit
- application created from multiple translation units, some TUs with NDEBUG macro undefined, others have NDEBUG defined
- application created from multiple TUs, all of them with NDEBUG undefined

Dependencies: all tests are expected to pass after merging the assert implementation in [3767](https://github.com/intel/llvm/pull/3767)

### CODE REVIEW

File: SYCL/Assert/assert_in_simultaneously_multiple_tus_one_ndebug.cpp [hunk 1]
```diff
@@ -0,0 +1,8 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple -DDEFINE_NDEBUG_INFILE2 -I %S/Inputs %S/assert_in_simultaneously_multiple_tus.cpp %S/Inputs/kernels_in_file2.cpp -o %t.out %threads_lib
+// RUN: %CPU_RUN_PLACEHOLDER %t.out | %CPU_RUN_PLACEHOLDER FileCheck %s
+// RUN: %GPU_RUN_PLACEHOLDER %t.out | %GPU_RUN_PLACEHOLDER FileCheck %s
+// RUN: %ACC_RUN_PLACEHOLDER %t.out | %ACC_RUN_PLACEHOLDER FileCheck %s
```
Comment (vladimirlaz) [inline SYCL/Assert/assert_in_simultaneously_multiple_tus_one_ndebug.cpp]:
```suggestion
// RUN: %CPU_RUN_PLACEHOLDER %t.out %CPU_CHECK_PLACEHOLDER
// RUN: %GPU_RUN_PLACEHOLDER %t.out %GPU_CHECK_PLACEHOLDER
// RUN: %ACC_RUN_PLACEHOLDER %t.out %ACC_CHECK_PLACEHOLDER
```

File: SYCL/Assert/assert_in_simultaneously_multiple_tus.cpp [hunk 2]
```diff
@@ -0,0 +1,87 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple -I %S/Inputs %s %S/Inputs/kernels_in_file2.cpp -o %t.out %threads_lib
+// RUN: %CPU_RUN_PLACEHOLDER %t.out | %CPU_RUN_PLACEHOLDER FileCheck %s
+// RUN: %GPU_RUN_PLACEHOLDER %t.out | %GPU_RUN_PLACEHOLDER FileCheck %s
+// RUN: %ACC_RUN_PLACEHOLDER %t.out | %ACC_RUN_PLACEHOLDER FileCheck %s
```
Comment (vladimirlaz) [inline SYCL/Assert/assert_in_simultaneously_multiple_tus.cpp]:
```suggestion
// RUN: %CPU_RUN_PLACEHOLDER %t.out %CPU_CHECK_PLACEHOLDER
// RUN: %GPU_RUN_PLACEHOLDER %t.out %GPU_CHECK_PLACEHOLDER
// RUN: %ACC_RUN_PLACEHOLDER %t.out %ACC_CHECK_PLACEHOLDER
```

File: SYCL/Assert/assert_in_simultaneous_kernels.cpp [hunk 3]
```diff
@@ -0,0 +1,74 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out %threads_lib
+// RUN: %CPU_RUN_PLACEHOLDER %t.out | %CPU_RUN_PLACEHOLDER FileCheck %s
+// RUN: %GPU_RUN_PLACEHOLDER %t.out | %GPU_RUN_PLACEHOLDER FileCheck %s
+// RUN: %ACC_RUN_PLACEHOLDER %t.out | %ACC_RUN_PLACEHOLDER FileCheck %s
```
Comment (vladimirlaz) [inline SYCL/Assert/assert_in_simultaneous_kernels.cpp]:
```suggestion
// RUN: %CPU_RUN_PLACEHOLDER %t.out %CPU_CHECK_PLACEHOLDER
// RUN: %GPU_RUN_PLACEHOLDER %t.out %GPU_CHECK_PLACEHOLDER
// RUN: %ACC_RUN_PLACEHOLDER %t.out %ACC_CHECK_PLACEHOLDER
```

File: SYCL/Assert/assert_in_one_kernel_ndebug.cpp [hunk 4]
```diff
@@ -0,0 +1,7 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple -DNDEBUG  %S/assert_in_one_kernel.cpp -o %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out | %CPU_RUN_PLACEHOLDER FileCheck %s
+// RUN: %GPU_RUN_PLACEHOLDER %t.out | %GPU_RUN_PLACEHOLDER FileCheck %s
+// RUN: %ACC_RUN_PLACEHOLDER %t.out | %ACC_RUN_PLACEHOLDER FileCheck %s
```
Comment (vladimirlaz) [inline SYCL/Assert/assert_in_one_kernel_ndebug.cpp]:
```suggestion
// RUN: %CPU_RUN_PLACEHOLDER %t.out %CPU_CHECK_PLACEHOLDER
// RUN: %GPU_RUN_PLACEHOLDER %t.out %GPU_CHECK_PLACEHOLDER
// RUN: %ACC_RUN_PLACEHOLDER %t.out %ACC_CHECK_PLACEHOLDER
```

File: SYCL/Assert/assert_in_one_kernel.cpp [hunk 5]
```diff
@@ -0,0 +1,43 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out | %CPU_RUN_PLACEHOLDER FileCheck %s
+// RUN: %GPU_RUN_PLACEHOLDER %t.out | %GPU_RUN_PLACEHOLDER FileCheck %s
+// RUN: %ACC_RUN_PLACEHOLDER %t.out | %ACC_RUN_PLACEHOLDER FileCheck %s
```
Comment (vladimirlaz) [inline SYCL/Assert/assert_in_one_kernel.cpp]:
```suggestion
// RUN: %CPU_RUN_PLACEHOLDER %t.out %CPU_CHECK_PLACEHOLDER
// RUN: %GPU_RUN_PLACEHOLDER %t.out %GPU_CHECK_PLACEHOLDER
// RUN: %ACC_RUN_PLACEHOLDER %t.out %ACC_CHECK_PLACEHOLDER
```

File: SYCL/Assert/assert_in_multiple_tus_one_ndebug.cpp [hunk 6]
```diff
@@ -0,0 +1,10 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple -DDEFINE_NDEBUG_INFILE2 -I %S/Inputs %S/assert_in_multiple_tus.cpp %S/Inputs/kernels_in_file2.cpp -o %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out | %CPU_RUN_PLACEHOLDER FileCheck %s
+// RUN: %GPU_RUN_PLACEHOLDER %t.out | %GPU_RUN_PLACEHOLDER FileCheck %s
+// RUN: %ACC_RUN_PLACEHOLDER %t.out | %ACC_RUN_PLACEHOLDER FileCheck %s
```
Comment (vladimirlaz) [inline SYCL/Assert/assert_in_multiple_tus_one_ndebug.cpp]:
```suggestion
// RUN: %CPU_RUN_PLACEHOLDER %t.out %CPU_CHECK_PLACEHOLDER
// RUN: %GPU_RUN_PLACEHOLDER %t.out %GPU_CHECK_PLACEHOLDER
// RUN: %ACC_RUN_PLACEHOLDER %t.out %ACC_CHECK_PLACEHOLDER
```

File: SYCL/Assert/assert_in_multiple_tus.cpp [hunk 7]
```diff
@@ -0,0 +1,60 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple -I %S/Inputs %s %S/Inputs/kernels_in_file2.cpp -o %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out | %CPU_RUN_PLACEHOLDER FileCheck %s
+// RUN: %GPU_RUN_PLACEHOLDER %t.out | %GPU_RUN_PLACEHOLDER FileCheck %s
+// RUN: %ACC_RUN_PLACEHOLDER %t.out | %ACC_RUN_PLACEHOLDER FileCheck %s
```
Comment (vladimirlaz) [inline SYCL/Assert/assert_in_multiple_tus.cpp]:
```suggestion
// RUN: %CPU_RUN_PLACEHOLDER %t.out %CPU_CHECK_PLACEHOLDER
// RUN: %GPU_RUN_PLACEHOLDER %t.out %GPU_CHECK_PLACEHOLDER
// RUN: %ACC_RUN_PLACEHOLDER %t.out %ACC_CHECK_PLACEHOLDER
```

File: SYCL/Assert/assert_in_kernels_ndebug.cpp [hunk 8]
```diff
@@ -0,0 +1,9 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple -DNDEBUG %S/assert_in_kernels.cpp -o %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out | %CPU_RUN_PLACEHOLDER FileCheck %s
+// RUN: %GPU_RUN_PLACEHOLDER %t.out | %GPU_RUN_PLACEHOLDER FileCheck %s
+// RUN: %ACC_RUN_PLACEHOLDER %t.out | %ACC_RUN_PLACEHOLDER FileCheck %s
```
Comment (vladimirlaz) [inline SYCL/Assert/assert_in_kernels_ndebug.cpp]:
```suggestion
// RUN: %CPU_RUN_PLACEHOLDER %t.out %CPU_CHECK_PLACEHOLDER
// RUN: %GPU_RUN_PLACEHOLDER %t.out %GPU_CHECK_PLACEHOLDER
// RUN: %ACC_RUN_PLACEHOLDER %t.out %ACC_CHECK_PLACEHOLDER
```

File: SYCL/Assert/assert_in_kernels.cpp [hunk 9]
```diff
@@ -0,0 +1,83 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out | %CPU_RUN_PLACEHOLDER FileCheck %s
+// RUN: %GPU_RUN_PLACEHOLDER %t.out | %GPU_RUN_PLACEHOLDER FileCheck %s
+// RUN: %ACC_RUN_PLACEHOLDER %t.out | %ACC_RUN_PLACEHOLDER FileCheck %s
```
Comment (vladimirlaz) [inline SYCL/Assert/assert_in_kernels.cpp]:
```suggestion
// RUN: %CPU_RUN_PLACEHOLDER %t.out %CPU_CHECK_PLACEHOLDER
// RUN: %GPU_RUN_PLACEHOLDER %t.out %GPU_CHECK_PLACEHOLDER
// RUN: %ACC_RUN_PLACEHOLDER %t.out %ACC_CHECK_PLACEHOLDER
```

File: SYCL/Assert/assert_in_simultaneously_multiple_tus_one_ndebug.cpp (hunk 1 above)
Comment (vladimirlaz) [inline SYCL/Assert/assert_in_simultaneously_multiple_tus_one_ndebug.cpp]:
```suggestion
// RUN: %CPU_RUN_PLACEHOLDER %t.out %CPU_CHECK_PLACEHOLDER
// RUN: %GPU_RUN_PLACEHOLDER %t.out %GPU_CHECK_PLACEHOLDER
// RUN: %ACC_RUN_PLACEHOLDER %t.out %ACC_CHECK_PLACEHOLDER
```

File: SYCL/Assert/assert_in_simultaneously_multiple_tus.cpp (hunk 2 above)
Comment (vladimirlaz) [inline SYCL/Assert/assert_in_simultaneously_multiple_tus.cpp]:
```suggestion
// RUN: %CPU_RUN_PLACEHOLDER %t.out %CPU_CHECK_PLACEHOLDER
// RUN: %GPU_RUN_PLACEHOLDER %t.out %GPU_CHECK_PLACEHOLDER
// RUN: %ACC_RUN_PLACEHOLDER %t.out %ACC_CHECK_PLACEHOLDER
```

File: SYCL/Assert/Inputs/kernels_in_file2.cpp [hunk 10]
```diff
@@ -0,0 +1,45 @@
+#include "kernels_in_file2.hpp"
```
Comment (vladimirlaz) [inline SYCL/Assert/Inputs/kernels_in_file2.cpp:1]: It is not needed here
```suggestion
```

File: SYCL/Assert/Inputs/kernels_in_file2.hpp [hunk 11]
```diff
@@ -0,0 +1,9 @@
+#include <CL/sycl.hpp>
+
+using namespace cl::sycl;
```
Comment (vladimirlaz) [inline SYCL/Assert/Inputs/kernels_in_file2.hpp]:
```suggestion
```

File: SYCL/Assert/assert_in_one_kernel_ndebug.cpp (hunk 4 above)
Comment (vladimirlaz) [inline SYCL/Assert/assert_in_one_kernel_ndebug.cpp]:
```suggestion
// RUN: %CPU_RUN_PLACEHOLDER %t.out %CPU_CHECK_PLACEHOLDER
// RUN: %GPU_RUN_PLACEHOLDER %t.out %GPU_CHECK_PLACEHOLDER
// RUN: %ACC_RUN_PLACEHOLDER %t.out %ACC_CHECK_PLACEHOLDER
```

**Review by vladimirlaz — CHANGES_REQUESTED** [review]

File: SYCL/Assert/Inputs/kernels_in_file2.hpp (hunk 11 above)
Comment (s-kanaev) [reply SYCL/Assert/Inputs/kernels_in_file2.hpp]: Removed

File: SYCL/Assert/assert_in_kernels.cpp (hunk 9 above)
Comment (s-kanaev) [reply SYCL/Assert/assert_in_kernels.cpp]: Fixed

File: SYCL/Assert/assert_in_kernels_ndebug.cpp (hunk 8 above)
Comment (s-kanaev) [reply SYCL/Assert/assert_in_kernels_ndebug.cpp]: Fixed.

File: SYCL/Assert/assert_in_multiple_tus.cpp (hunk 7 above)
Comment (s-kanaev) [reply SYCL/Assert/assert_in_multiple_tus.cpp]: Fixed.

File: SYCL/Assert/assert_in_multiple_tus_one_ndebug.cpp (hunk 6 above)
Comment (s-kanaev) [reply SYCL/Assert/assert_in_multiple_tus_one_ndebug.cpp]: Fixed

File: SYCL/Assert/assert_in_one_kernel.cpp (hunk 5 above)
Comment (s-kanaev) [reply SYCL/Assert/assert_in_one_kernel.cpp]: Fixed

File: SYCL/Assert/assert_in_one_kernel_ndebug.cpp (hunk 4 above)
Comment (s-kanaev) [reply SYCL/Assert/assert_in_one_kernel_ndebug.cpp]: Fixed.

File: SYCL/Assert/assert_in_simultaneous_kernels.cpp (hunk 3 above)
Comment (s-kanaev) [reply SYCL/Assert/assert_in_simultaneous_kernels.cpp]: Fixed.

File: SYCL/Assert/assert_in_simultaneously_multiple_tus.cpp (hunk 2 above)
Comment (s-kanaev) [reply SYCL/Assert/assert_in_simultaneously_multiple_tus.cpp]: Fixed

File: SYCL/Assert/assert_in_simultaneously_multiple_tus_one_ndebug.cpp (hunk 1 above)
Comment (s-kanaev) [reply SYCL/Assert/assert_in_simultaneously_multiple_tus_one_ndebug.cpp]: Fixed.

**Review by vzakhari — DISMISSED** [review]
DeviceLib change LGTM.

**Review by vladimirlaz — DISMISSED** [review]

**Comment by bader** [comment]
@alexanderfle, I've updated code owners for Assert tests.

**Review by vzakhari — DISMISSED** [review]

**Review by vladimirlaz — DISMISSED** [review]

File: SYCL/Assert/assert_in_simultaneously_multiple_tus.cpp [hunk 12]
```diff
@@ -0,0 +1,96 @@
+// FIXME unsuppoerted on CUDA until fallback libdevice becomes available
+// UNSUPPORTED: cuda
+// FIXME disable the test on Linux on OpenCL on FPGA as there's some funny
+// failure taking place in this configuration on CI. The test should be enabled
+// with this configuration once it's debugged.
+// UNSUPPORTED: linux && opencl && fpga
```
Comment (s-kanaev) [inline SYCL/Assert/assert_in_simultaneously_multiple_tus.cpp:6]: @vladimirlaz does this seem correct way to disable the test only on Linux with OpenCL for FPGA device?

File: SYCL/Assert/assert_in_simultaneously_multiple_tus_one_ndebug.cpp [hunk 13]
```diff
@@ -0,0 +1,17 @@
+// FIXME unsuppoerted on CUDA until fallback libdevice becomes available
+// UNSUPPORTED: cuda
+// FIXME disable the test on Linux on OpenCL on FPGA as there's some funny
+// failure taking place in this configuration on CI. The test should be enabled
+// with this configuration once it's debugged.
+// UNSUPPORTED: linux && opencl && fpga
```
Comment (s-kanaev) [inline SYCL/Assert/assert_in_simultaneously_multiple_tus_one_ndebug.cpp:6]: @vladimirlaz does this seem correct way to disable the test only on Linux with OpenCL for FPGA device?

Comment (vladimirlaz) [reply SYCL/Assert/assert_in_simultaneously_multiple_tus_one_ndebug.cpp:6]: yes

File: SYCL/Assert/assert_in_simultaneously_multiple_tus.cpp (hunk 12 above)
Comment (vladimirlaz) [reply SYCL/Assert/assert_in_simultaneously_multiple_tus.cpp:6]: yes

**Review by vladimirlaz — APPROVED** [review]

**Comment by s-kanaev** [comment]
The sporadic failure of L0 batching test was reported.

@vzakhari , @bader , @romanovvlad , a friendly ping on code-review

**Comment by romanovvlad** [comment]
> 
> 
> The sporadic failure of L0 batching test was reported.
> 
> @vzakhari , @bader , @romanovvlad , a friendly ping on code-review

I believe Vladimir's review should be enough from the RT side.

**Review by vzakhari — APPROVED** [review]

**Pull request merged**